synapse/handlers/oidc_handler.py

```diff
@@ -674,38 +674,43 @@ async def handle_oidc_callback(self, request: SynapseRequest) -> None:
                 self._sso_handler.render_error(request, "invalid_token", str(e))
                 return
 
-        # Pull out the user-agent and IP from the request.
-        user_agent = request.get_user_agent("")
-        ip_address = self.hs.get_ip_from_request(request)
-
-        # Call the mapper to register/login the user
         try:
+            # first check if we're doing a UIA
+            if ui_auth_session_id:
+                return await self._sso_handler.complete_sso_ui_auth_request(
+                    self._auth_provider_id,
+                    self._remote_id_from_userinfo(userinfo),
+                    ui_auth_session_id,
+                    request,
+                )
+
+            # otherwise, it's a login
+
+            # Pull out the user-agent and IP from the request.
+            user_agent = request.get_user_agent("")
+            ip_address = self.hs.get_ip_from_request(request)
+
+            # Call the mapper to register/login the user
             user_id = await self._map_userinfo_to_user(
                 userinfo, token, user_agent, ip_address
             )
-        except MappingException as e:
-            logger.exception("Could not map user")
-            self._sso_handler.render_error(request, "mapping_error", str(e))
-            return
 
-        # Mapping providers might not have get_extra_attributes: only call this
-        # method if it exists.
-        extra_attributes = None
-        get_extra_attributes = getattr(
-            self._user_mapping_provider, "get_extra_attributes", None
-        )
-        if get_extra_attributes:
-            extra_attributes = await get_extra_attributes(userinfo, token)
-
-        # and finally complete the login
-        if ui_auth_session_id:
-            await self._auth_handler.complete_sso_ui_auth(
-                user_id, ui_auth_session_id, request
+            # Mapping providers might not have get_extra_attributes: only call this
+            # method if it exists.
+            extra_attributes = None
+            get_extra_attributes = getattr(
+                self._user_mapping_provider, "get_extra_attributes", None
             )
-        else:
+            if get_extra_attributes:
+                extra_attributes = await get_extra_attributes(userinfo, token)
+
+            # and finally complete the login
             await self._auth_handler.complete_sso_login(
                 user_id, request, client_redirect_url, extra_attributes
             )
+        except MappingException as e:
+            logger.exception("Could not map user")
+            self._sso_handler.render_error(request, "mapping_error", str(e))
```

Comment:
Does it make sense to push this try-except across the whole function like this? We don't really expect most of this code to raise exceptions and it is usually bad-practice to over-expand try-blocks.

(The same goes for the SAML code.)

Reply:
maybe not, actually. I can't quite remember why I did this. I'll tighten it up.